src/test/wwidgetstack_test.cpp

```diff
@@ -0,0 +1,207 @@
+            EXPECT_TRUE(m_pPage1Widget->isHidden());
+            EXPECT_FALSE(m_pPage2Widget->isHidden());
+            break;
+        }
+    }
+
+    QScopedPointer<WWidgetStack> m_pStack;
+    QScopedPointer<ControlObject> m_pPrevControl;
+    QScopedPointer<ControlObject> m_pNextControl;
+    QScopedPointer<ControlObject> m_pCurPageControl;
+    QScopedPointer<QWidget> m_pPage0Widget;
+    QScopedPointer<QWidget> m_pPage1Widget;
+    QScopedPointer<QWidget> m_pPage2Widget;
+    QScopedPointer<ControlObject> m_pPage0Control;
+    QScopedPointer<ControlObject> m_pPage1Control;
+    QScopedPointer<ControlObject> m_pPage2Control;
+    const char* m_pGroup;
+};
+
+TEST_F(WWidgetStackTest, MaintainPageSelected) {
+    // As a widget stack is created, if the current page control object
+    // is already set, that setting should remain the same and the final stack
+    // should be consistent.  Since we set the control in SetUp, that part
+    // is already done.  We just need to test the current state.
+    ExpectPageSelected(1);
+}
+
+TEST_F(WWidgetStackTest, MaintainPageControlValue) {
+    // The current page control overrides whatever values the individual
+    // page triggers may already have.  This is what caused the bug with the
+    // LateNight skin -- The page trigger was set to off, but the current page
+    // control was defaulted to 0, so that overrode the page trigger and
+    // showed the first page.
+
+    // This test is set up to reproduce the original LateNight skin case.
+    m_pCurPageControl.reset(
+            new ControlObject(ConfigKey(m_pGroup,
+                                        "MaintainPageControlValue-page")));
+    QScopedPointer<WWidgetStack> stack(
+            new WWidgetStack(NULL, m_pNextControl.data(),
+                             m_pPrevControl.data(),
+                             m_pCurPageControl.data()));
+
+    QWidget page0;
+    QWidget page1;
+
+    // We don't set the current page control.  All we know is that page0
+    // is supposed to be off.
+    m_pPage0Control->set(0);
+
+    stack->addWidgetWithControl(&page0, m_pPage0Control.data(), -1);
+    stack->addWidgetWithControl(&page1, NULL, -1);
+    stack->Init();
+
+    // The off state above is overridden by the default value of curpagecontrol,
+    // which is set to 0 on creation.
+    ExpectPageSelected(0);
+}
+
+TEST_F(WWidgetStackTest, ChangePageSelection) {
+    // Changing the page via the current page control should work.
+    m_pCurPageControl->set(2);
+    ExpectPageSelected(2);
+    m_pCurPageControl->set(0);
+    ExpectPageSelected(0);
+}
+
+TEST_F(WWidgetStackTest, ChangeChildControls) {
+    // Changing the page via the individual page controls should work.
+    m_pPage0Control->set(1);
+    ExpectPageSelected(0);
+    m_pPage2Control->set(1);
+    ExpectPageSelected(2);
+}
+
+TEST_F(WWidgetStackTest, OnHideBehavior) {
+    // When hiding a page, if the on_hide_select value is -1, the next
```

Comment:
>>   flaky

> I don't see how it could be flaky. If I do ./mixxx-test --gtest_repeat=1000 it does not fail.

That is not the problem here. It is flaky, because it fixes an undesired behavior. Once the TODO it is fixed this test will fail.